ouroboros-network-framework/test/Test/Simulation/Network/Snocket.hs

```diff
@@ -627,28 +638,208 @@ prop_generator_NonFailingBeararInfoScript (NonFailingBearerInfoScript s) = not (
+  let tr = runSimTrace sim
+   in case traceResult True tr of
+     Left e -> counterexample
+                (unlines
+                  [ "=== Say Events ==="
+                  , unlines (selectTraceEventsSay' tr)
+                  , "=== Trace Events ==="
+                  , unlines (show `map` traceEvents tr)
+                  , "=== Error ==="
+                  , show e ++ "\n"
+                  ])
+                False
+     Right (Left e) -> counterexample
+                        (unlines
+                          [ "=== Say Events ==="
+                          , unlines (selectTraceEventsSay' tr)
+                          , "=== Trace Events ==="
+                          , unlines (show `map` traceEvents tr)
+                          , "=== Error ==="
+                          , show e ++ "\n"
+                          ])
+                        False
+     Right (Right _) -> property True
+
+prop_connect_to_accepting_socket :: BearerInfoScript -> Property
+prop_connect_to_accepting_socket (BearerInfoScript script) =
+    prop_verify_connect_to_simulation sim
+  where
+    sim :: forall s . IOSim s (Either SomeException ())
+    sim =
+      withSnocket nullTracer (toBearerInfo <$> script) $ \snocket ->
+        withAsync
+          (runServer (TestAddress (0 :: Int)) snocket (close snocket) acceptOne)
+          $ \serverAsync -> do
+            res <- runClient (TestAddress 1) (TestAddress 0)
+                            snocket (close snocket)
+            _ <- wait serverAsync
+            return res
+
+prop_connect_and_not_close :: BearerInfoScript -> Property
+prop_connect_and_not_close (BearerInfoScript script) =
+    prop_verify_connect_to_simulation sim
+  where
+    sim :: forall s . IOSim s (Either SomeException ())
+    sim =
+      withSnocket nullTracer (toBearerInfo <$> script) (\snocket ->
+        withAsync
+          (runServer (TestAddress (0 :: Int)) snocket (\_ -> pure ()) acceptOne)
+          $ \serverAsync -> do
+            res <- runClient (TestAddress 1) (TestAddress 0)
+                            snocket (\_ -> pure ())
+            _ <- wait serverAsync
+            return res
+                                                       )
+      `catch` \(err :: SomeException) ->
+        -- Should error with NotReleasedListeningSockets
+        case (fromException err
+               :: Maybe (ResourceException (TestAddress Int))) of
+          Just _ ->
+            return (Right ())
+          Nothing ->
+            return (Left err)
+
+prop_connect_to_not_accepting_socket :: BearerInfoScript -> Property
+prop_connect_to_not_accepting_socket (BearerInfoScript script) =
+    prop_verify_connect_to_simulation sim
+  where
+    sim :: forall s . IOSim s (Either SomeException ())
+    sim =
+      withSnocket nullTracer (toBearerInfo <$> script) $ \snocket ->
+        withAsync (runServer (TestAddress (0 :: Int)) snocket
+                             (close snocket) loop)
```

Comment:
`loop` seems to be only used in this function, I would move it the `where` close here.